Honor LogAppendTime when decoding messages

hrubi opened the pull request:
When LogAppendTime is in use, MaxTimestamp should be set as the
timestamp for all Records in the RecordBatch.

hrubi committed:
Honor LogAppendTime when decoding messages

For messages format version 1:

When LogAppendTime is in use and compression is enabled on the wrapper
message, set the timestamp of each inner message to the timestamp of the
wrapper message.

This is documented in KIP-32 [1].

For messages format version 2:

When LogAppendTime is in use, set MaxTimestamp as the timestamp for all
Records in the RecordBatch. FirstTimestamp + TsDelta always form the
CreateTime, which is unwanted in this case. There's no explicit mention
in the Kafka documentation. See the official Java client implementation
[2], [3].

The fetch test now uses additional topic with the LogAppendTime
configuration enabled. The test checks that the log append time returned
when producing a message batch matches the timestamp in the fetched
messages. It also checks that the create timestamp set by the producer
is different from the timestamp in the fetched message. The create
timestamp is lowered by by a random number to guarantee that it's in the
past and that Kafka's log append time is greater. Randomizing the
timestamp also makes the test which checks the create timestamp more
robust.

[1] https://cwiki.apache.org/confluence/display/KAFKA/KIP-32+-+Add+timestamps+to+Kafka+message#KIP-32-AddtimestampstoKafkamessage-AddaTimestampfieldtothemessageformatwithmaximumallowedtimedifferenceconfigurationonbroker.
[2] https://github.com/apache/kafka/blob/1f1179ea64bbaf068d759aae988bd2a6fe966161/clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java#L558
[3] https://github.com/apache/kafka/blob/1f1179ea64bbaf068d759aae988bd2a6fe966161/clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java#L330-L331

hrubi commented:
I've sent a mail to the Kafka mailing list: https://lists.apache.org/thread.html/bc98bc5365a4388d799bf9834b1dfff56c7d03b19f29c48f07abe1ad@%3Cusers.kafka.apache.org%3E, requesting clarification of the expected client behaviour.

Here's the current behaviour of the official Java client:
* [DefaultRecordBatch.java](https://github.com/apache/kafka/blob/1f1179ea64bbaf068d759aae988bd2a6fe966161/clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java#L558)
* [DefaultRecord.java](https://github.com/apache/kafka/blob/1f1179ea64bbaf068d759aae988bd2a6fe966161/clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java#L330-L331)

I've also find the same in the [librdkafka](https://github.com/edenhill/librdkafka/blob/3cc2db5ae5177604223748cd92dbc9089d8a2e40/src/rdkafka_msgset_reader.c#L894-L909) library. Unfortunately, I haven't find this in any docs.

Also, I should likely write a test for this. What do you think?

zmstone commented:
thanks for the PR and all the investigations.
was it not 0 sent by kafka ? if no, what is it ?

zmstone commented:
there should be a version bump in .src file.

zmstone commented:
about tests, maybe create a new topic with `message.timestamp.type` configured in this script: https://github.com/klarna/kafka_protocol/blob/master/scripts/setup-testbed.sh
then add a test to https://github.com/klarna/kafka_protocol/blob/master/test/kpro_fetch_tests.erl

hrubi commented:
Thanks for a quick response.

In the record batch, the FirstTimestamp is set to the CreateTime of the first record in the batch, so kafka_protocol was always returning the CreateTime, even when LogAppendTime was used. I will describe it in more detail in the commit message. I also wanted to link to relevant docs, or at least to a response to my inquiry on the Kafka ML once I receive it.

I'll bump the .src and create a test.

zmstone commented:
> was it not 0 sent by kafka ? if no, what is it ?

Sorry, I was not quite clear in this question, I meant to ask if it was 0 for the `delta`.
But you gave the desired answer anyway. thanks.

hrubi commented:
I've just tried this (from Elixir's IEx):
```
:brod.produce_sync(:my_client, "lat_uncompressed", 0, "", [%{ts: 1234, value: "a"}, %{ts: 1235, value: "b"}, %{ts: 1400, value: "c"}])
```

And when I print out few variables in kpro_batch during `:brod_utils.fetch`, I get in `do_decode`:
FirstTimestamp: 1234
MaxTimestamp: 1570630122297

In `dec_record`:
TsDelta "a": 0
TsDelta "b": 1
TsDelta "c": 166

So the FirstTimestamp with the TsDelta is the CreateTime of each record.

hrubi committed:
Remove extraneous function clauses in test

The third argument to `validate_message` is a map created in
`make_random_batch`, so it can never be a tuple.

hrubi committed:
Assert timestamps in fetch test

Check that the timestamp of the produced and fetched message matches. It
is relevant only when the message is produced with produce API >= 1 and
fetched with fetch API >= 2.

hrubi commented:
I've pushed the following.

2dfc428:
* Minor cleanup.

a5bc259:
* Extended fetch tests with timestamp checking for the current behaviour, as it made sense to me to check the create timestamps when I'm going to check the append timestamps next.

d93375d:
* Coding style whitespace fix in the patch in src/kpro_batch.erl, besides that it's the same as before.
* Test for the log append time.

hrubi commented on test/kpro_fetch_tests.erl:
```diff
@@ -81,21 +88,45 @@ drop_older_offsets(Offset, [#kafka_message{offset = O} | R] = ML) ->
+                , magic => Magic
+                , ts_type => TType
+                , log_append_time => LAT
+                }).
```
This should be a bit different from the `validate_message_create`, because if the timestamp is present depends solely on the fetch api version. Also it depends on the producer api version used if we do or do not have the LAT available.

So the clauses could be as follows:
* `FetchVsn < 2` → no timestamp check
* `ProduceVsn < 2` → no LAT, check only produce timestamp differs from the fetched timestamp
* Anything else → check both LAT and produce/fetch timestamp difference.

Another idea is to check that `LogAppendTime` is returned in `produce_randomly` when the produce api vsn is >= 2. The `validate_message_append` will be just slightly changed and the `Magic` would depend just on the fetch api version.

It's getting more and more complicated, I will think about how to make the tests more readable.

hrubi replied:
To simplify this a lot, it could go in the opposite direction - assert based just on the timestamp type in the fetched messages (and don't care about the api versions during assertions).

The `validate_message` then could look like this:
```erlang
% message magic 0, no timestamp
validate_message(undefined, _T, K, V, #{key := K, value := V}) -> ok;
% message produced without timestamp
validate_message(create, -1, K, V, #{key := K, value := V}) -> ok;
% message produced with matching create timestamp
validate_message(create, T, K, V, #{ts := T, key := K, value := V}) -> ok;
% message produced with create timestamp which differs from the fetched log append timestamp
validate_message(append, TF, K, V, #{ts := TP, key := K, value := V}) when TF /= TP -> ok;
% something unexpected
validate_message(TType, T, K, V, Wat) ->
  erlang:error(#{ fetched => {TType, T, K, V}
                , produced => Wat
                }).
```

hrubi replied:
I've pushed the simple approach here: https://github.com/hrubi/kafka_protocol/commits/decode-log-append-time-simple-tests

@k32 Thanks for taking care of it. Please let me know what do you think about the tests.

zmstone commented:
I feel like I can't be responsive in this thread.
@k32 can help out?

k32 commented:
Hi! Thanks for the PR and for the reviews. I'll need some time to digest all the comments.

k32 commented on test/kpro_fetch_tests.erl:
```diff
@@ -81,21 +88,45 @@ drop_older_offsets(Offset, [#kafka_message{offset = O} | R] = ML) ->
 
-do_fetch(Connection, Vsn, BeginOffset, MaxBytes) ->
-  Req = make_req(Vsn, BeginOffset, MaxBytes),
+validate_message_append(_Magic = 0, _TType, _T, K, V, _LAT, #{key := K, value := V}) -> ok;
```
This check is getting pretty hard to understand with all these new parameters and single- and two-letter variables. Suggestion: pack all the test and reference data into maps, and match only fields you need?

k32 commented:
Hi @hrubi ! Sorry for not responding. This PR looks good to me in general, apart from small remarks. But the new tests fail in CI: https://travis-ci.org/klarna/kafka_protocol/jobs/595867668 Could you please take a look.

Due to org-wide changes in the repository security settings Travis stopped posting test results to the PRs, I am trying to fix it.

hrubi commented:
I've updated the following:
* To make the tests more readable, I've split asserting key/value - which is common for all code paths - and asserting timestamp - which is different for the append and create scenarios.
* I've also used maps instead of many cryptic variable names, as you have suggested.
* Thanks a lot for pointing out the failing build. It has revealed a similar bug in the `kpro_batch_v01`. Patching it fixes the failing test with Kafka 0.10.
* The other failing test with Kafka 0.9 was understandable, because that version does not support timestamps yet, so it was failing on the topic creation with LAT setting. The LAT test is now skipped altogether in that case. The `validate_message_ts_create` passes as it matches the low produce/fetch APIs, so it skips the timestamp check.

k32 commented:
Sorry for the silence, your PR isn't forgotten! I just need more time to analyze it. I will respond during weekend.

k32 commented:
Thanks for the PR!

hrubi commented:
Thanks a lof for taking care of it!
